src/mscorlib/corefx/System/Globalization/CalendarData.Unix.cs

```diff
@@ -1,47 +1,363 @@
+
+            return index - startIndex;
+        }
+
+        private static bool EnumMonthNames(string localeName, CalendarId calendarId, CalendarDataType dataType, out string[] monthNames)
+        {
+            monthNames = null;
+
+            CallbackContext callbackContext = new CallbackContext();
+            bool result = EnumCalendarInfo(localeName, calendarId, dataType, callbackContext);
+            if (result)
+            {
+                // the month-name arrays are expected to have 13 elements.  If ICU only returns 12, add an
+                // extra empty string to fill the array.
+                if (callbackContext.Results.Count == 12)
+                {
+                    callbackContext.Results.Add(string.Empty);
+                }
+
+                monthNames = callbackContext.Results.ToArray();
+            }
+
+            return result;
+        }
+
+        private static bool EnumEraNames(string localeName, CalendarId calendarId, CalendarDataType dataType, out string[] eraNames)
+        {
+            bool result = EnumCalendarInfo(localeName, calendarId, dataType, out eraNames);
+
+            // .NET expects that only the Japanese calendars have more than 1 era.
+            // So for other calendars, only return the latest era.
+            if (calendarId != CalendarId.JAPAN && calendarId != CalendarId.JAPANESELUNISOLAR && eraNames.Length > 0)
+            {
+                string[] latestEraName = new string[] { eraNames[eraNames.Length - 1] };
+                eraNames = latestEraName;
+            }
+
+            return result;
+        }
+
+        internal static bool EnumCalendarInfo(string localeName, CalendarId calendarId, CalendarDataType dataType, out string[] calendarData)
+        {
+            calendarData = null;
+
+            CallbackContext callbackContext = new CallbackContext();
+            bool result = EnumCalendarInfo(localeName, calendarId, dataType, callbackContext);
+            if (result)
+            {
+                calendarData = callbackContext.Results.ToArray();
+            }
+
+            return result;
+        }
+
+        private static bool EnumCalendarInfo(string localeName, CalendarId calendarId, CalendarDataType dataType, CallbackContext callbackContext)
+        {
+            GCHandle context = GCHandle.Alloc(callbackContext);
+            try
+            {
+                return Interop.GlobalizationInterop.EnumCalendarInfo(EnumCalendarInfoCallback, localeName, calendarId, dataType, (IntPtr)context);
+            }
+            finally
+            {
+                context.Free();
+            }
+        }
+
+        private static void EnumCalendarInfoCallback(string calendarString, IntPtr context)
+        {
+            CallbackContext callbackContext = (CallbackContext)((GCHandle)context).Target;
+
+            if (callbackContext.DisallowDuplicates)
+            {
+                foreach (string existingResult in callbackContext.Results)
```

Comment:
We expect the set of results to always be relatively small?  Otherwise this O(N^2) could get costly.